Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+    go f = case f of
+      Add x y g -> do
+        var <- lift freshVar
+        tell $
+          var <> " <- add " <> x <> " " <> y <> "\n"
+        pure $ g var
+      Input x -> do
+        var <- lift freshVar
+        tell $ var <> " <- input\n"
+        pure $ x var
+      Output s next -> do
+        tell $ "output " <> s <> "\n"
+        pure next
+```
+
+Notice, up to this point, all constructors of our `ASTF` had exactly one parameter that had anything to do with the functor.
+A question you might have now is what happens if we have several.
+If you gained a bit of intuition for free monads by now, you might suspect we get a branching computation, and that's exactly right!
+Let's add simple branching to our toy language to see how it works.
+
+```haskell
+data ASTF t a
+  = Add t t (t -> a)
+  | Input (t -> a)
+  | Output t a
+  | If t a a
+  deriving Functor
+```
+
+We add a new constructor, `If`, with one scalar parameter and two branches.
+We can now define a helper,
+
+```haskell
+if' :: t -> FreeAST t a -> FreeAST t a -> FreeAST t a
+if' cond t f = Free $ If cond t f
+```
+
+Notice how instead of `liftF` we have to use `Free` directly here -- our branches are already of type `FreeAST`.
+We can now modify our interpreter `computeAST` by adding another case:
+
+```haskell
+computeAST :: FreeAST Int () -> IO ()
+computeAST = foldFree go
+  where
+    go :: ASTF Int x -> IO x
+    go arg = case arg of
+      Add x y next -> pure $ next (x + y)
+      Input next -> next . read <$> getLine
+      Output x next -> do
+        print x
+        pure next
+      -- below is the only thing that changed
+      If cond t f ->
+        if cond /= 0 then pure t else pure f
+```
+
+Here we define the semantics of the branching -- for the sake of not complicating the example, we interpret `0` as false, and non-zero as true.
+
+When we try modifying the pretty printer, we run into a bit of a problem.
+Since our branching operation introduces a bit of non-determinism into the computation, our approach with `WriterT (State Int)` doesn't work anymore.
+One could fix it by explicitly adding some kind of non-determinism transformer on top of this, like `ListT` or `LogicT`, but let us instead write the pretty-printer directly.
+
+```haskell
+printAST :: FreeAST String a -> String
+printAST fast = snd $ evalState (runWriterT $ go fast) 0
+  where
+    freshVar :: State Int String
+    freshVar = do
+      n <- get
+      put $ n + 1
+      pure $ "x" <> show n
+    go :: FreeAST String a -> WriterT String (State Int) ()
+    go (Pure _) = pure ()
+    go (Free f) = case f of
+      Add x y g -> do
+        var <- lift freshVar
+        tell $
+          var <> " <- add " <> x <> " " <> y <> "\n"
+        go $ g var
+      Input x -> do
+        var <- lift freshVar
+        tell $ var <> " <- input\n"
+        go $ x var
+      Output s next -> do
+        tell $ "output " <> s <> "\n"
+        go next
+      If cond tru fls -> do
```

Comment:
**Nitpick**: why not just `true` and `false` or `onTrue`, `onFalse`?

Reply:
Good point, `onTrue` and `onFalse` looks more self-explanatory.

If you care to know the reason why I went with `tru` and `fls`, that's something I've dragged up (not necessarily intentionally) from pure untyped lambda calculus, where the Church encoding of booleans commonly uses these identifiers.

Reply:
> from pure untyped lambda calculus, where the Church encoding of booleans commonly uses these identifiers.

Ah, I see 👌